### PR TITLE
Added protector to yara (FreeRASP)

### PR DESCRIPTION
**Pull request by Fare9**
Added a protector to yara rules for APK.

* URL of the protector: [Talsec](https://www.talsec.app/freerasp-in-app-protection-security-talsec).
* Tested APK: [2b8faa038bf34474075a56e2fda7887a7df9c3c57db8a9f25547dc9374137ec9](https://www.virustotal.com/gui/file/2b8faa038bf34474075a56e2fda7887a7df9c3c57db8a9f25547dc9374137ec9).
* More information: [https://github.com/talsec/Free-RASP-Android](https://github.com/talsec/Free-RASP-Android)

### CODE REVIEW

File: apkid/rules/apk/protectors.yara [hunk 1]
```diff
@@ -88,3 +88,21 @@ rule vkey_apk : protector
     1 of ($asseta*) and
     1 of ($assetb*)
 }
+
+rule free_rasp : protector
+{
+  meta:
+    description = "FreeRASP Android Protector"
```
Comment (enovella) [inline apkid/rules/apk/protectors.yara:95]: What about only `FreeRASP`? Reasons:
- APKiD is only for Android
- Protector is already defined in the tag

File: apkid/rules/apk/protectors.yara [hunk 2]
```diff
@@ -88,3 +88,21 @@ rule vkey_apk : protector
...
+    url         = "https://www.talsec.app/freerasp-in-app-protection-security-talsec"
+    sample      = "2b8faa038bf34474075a56e2fda7887a7df9c3c57db8a9f25547dc9374137ec9"
```
Comment (enovella) [inline apkid/rules/apk/protectors.yara:97]: Don't you want to include your name on the `author` tag?

File: apkid/rules/apk/protectors.yara [hunk 3]
```diff
@@ -88,3 +88,21 @@ rule vkey_apk : protector
...
+  
+  strings:
+    $lib1   = /lib\/(x86\_64|armeabi\-v7a|arm64\-v8a|x86)\/libsecurity\.so/
+    $lib2   = /lib\/(x86\_64|armeabi\-v7a|arm64\-v8a|x86)\/libpolarssl\.so/
+    $asset1 = "assets/talsec"
```
Comment (enovella) [inline apkid/rules/apk/protectors.yara:102]: only 1 asset, why do we need to specify `asset1`? Better to simplify names

File: apkid/rules/apk/protectors.yara [hunk 4]
```diff
@@ -88,3 +88,21 @@ rule vkey_apk : protector
...
+
+rule free_rasp : protector
+{
+  meta:
+    description = "FreeRASP Android Protector"
+    url         = "https://www.talsec.app/freerasp-in-app-protection-security-talsec"
+    sample      = "2b8faa038bf34474075a56e2fda7887a7df9c3c57db8a9f25547dc9374137ec9"
+  
+  strings:
+    $lib1   = /lib\/(x86\_64|armeabi\-v7a|arm64\-v8a|x86)\/libsecurity\.so/
```
Comment (enovella) [inline apkid/rules/apk/protectors.yara:100]: what about `/lib\/(arm.*|x86.*)\/` to simplify things with the same match?

File: apkid/rules/apk/protectors.yara (hunk 2 above)
Comment (Fare9) [reply apkid/rules/apk/protectors.yara:97]: Added with new commit

**Comment by Fare9** [comment]
Added improvements of readability!

**Comment by apkunpacker** [comment]
> Added a protector to yara rules for APK.
> 
> * URL of the protector: [Talsec](https://www.talsec.app/freerasp-in-app-protection-security-talsec).
> * Tested APK: [2b8faa038bf34474075a56e2fda7887a7df9c3c57db8a9f25547dc9374137ec9](https://www.virustotal.com/gui/file/2b8faa038bf34474075a56e2fda7887a7df9c3c57db8a9f25547dc9374137ec9).
> * More information: https://github.com/talsec/Free-RASP-Android

@Fare9  This rule need slight modification as https://virustotal.com/gui/file/e10b8772fd9b6aaf8ba030c5bcb324fb9b91f34e893a62bdf238629df856e047
not matching with it . There is no asset/talsec file

**Comment by enovella** [comment]
Ready to be merged @Fare9. What do you think about @apkunpacker's comment?

File: apkid/rules/apk/protectors.yara [hunk 5]
```diff
@@ -88,3 +88,20 @@ rule vkey_apk : protector
...
+  meta:
+    description = "FreeRASP"
+    url         = "https://www.talsec.app/freerasp-in-app-protection-security-talsec"
+    sample      = "2b8faa038bf34474075a56e2fda7887a7df9c3c57db8a9f25547dc9374137ec9"
+    author      = "Fare9"
+
+  strings:
+    $lib1   = /lib\/(arm.*|x86.*)\/libsecurity\.so/
+    $lib2   = /lib\/(arm.*|x86.*)\/libpolarssl\.so/
+    $asset = "assets/talsec"
```
Comment (enovella) [inline apkid/rules/apk/protectors.yara:103]: @Fare9 can you balance the whitespaces?

**Comment by Fare9** [comment]
I can balance the whitespaces, regarding what @apkunpacker says, I have not been able to download the APK for testing it, I tested it with applications generated with Android Studio, and using the provided configuration from the FreeRASP webpage. In the cases I generated they had the asset of talsec, and also the libraries. The libsecurity.so, for example, is where all the protections of anti-analysis are implemented.

**Comment by apkunpacker** [comment]
> I can balance the whitespaces, regarding what @apkunpacker says, I have not been able to download the APK for testing it, I tested it with applications generated with Android Studio, and using the provided configuration from the FreeRASP webpage. In the cases I generated they had the asset of talsec, and also the libraries. The libsecurity.so, for example, is where all the protections of anti-analysis are implemented.

Here is sample attached
[TalsecRaspSample.apk.zip](https://github.com/rednaga/APKiD/files/9626165/TalsecRaspSample.apk.zip)

**Comment by Fare9** [comment]
Will check the yara with that apk, and modify it accordingly to detect that one, and detect all those from my dataset too as test. Thanks!

**Comment by Fare9** [comment]
For what I see there are not assets in this APK, but the libraries are included, I can change it tonight. And probably add some of DEX or ELF that they have in common between these apps.

File: apkid/rules/apk/protectors.yara [hunk 6]
```diff
@@ -88,3 +88,19 @@ rule vkey_apk : protector
...
+{
+  meta:
+    description = "FreeRASP"
+    url         = "https://www.talsec.app/freerasp-in-app-protection-security-talsec"
+    sample      = "2b8faa038bf34474075a56e2fda7887a7df9c3c57db8a9f25547dc9374137ec9"
+    author      = "Fare9"
+
+  strings:
+    $lib1   = /lib\/(arm.*|x86.*)\/libsecurity\.so/
+    $lib2   = /lib\/(arm.*|x86.*)\/libpolarssl\.so/
```
Comment (enovella) [inline apkid/rules/apk/protectors.yara:102]: Don't you like the idea of having the `talsec` asset? If this file is not in all the APKs, we can always create two rules to differentiate compilers or versions.

File: apkid/rules/dex/protectors.yara [hunk 7]
```diff
@@ -111,3 +111,32 @@ rule appdome_dex : protector
...
+    description = "FreeRASP"
+    sample      = "e10b8772fd9b6aaf8ba030c5bcb324fb9b91f34e893a62bdf238629df856e047"
+    url         = "https://www.talsec.app/freerasp-in-app-protection-security-talsec"
+    author      = "Fare9"
+  
+  strings:
+    // Decryption method found in DEX files, since strings will change
+    // and other offsets change, we add ?? to some instructions
+    $decryption = {
+      6e 10 ?? ?? 08 00 0a 00 db 00 00 02 23 01 ?? ??
```
Comment (enovella) [inline apkid/rules/dex/protectors.yara:127]: Would it be possible to display commented offsets this way?  This is only to make it clear to read in the future by others.
```sh
    // Most of some kind of runtime decryption method, signature = a(IIZI[I[[I[I)V
    $decrypt_method = {
      12 01                 // const/4 v1, 0x0
      39 ?? ?? ??           // if-nez ??, :????
      71 ?? ?? ?? ?? ??     // invoke-static {??}, ??
      01 10                 // move v0, v1
      35 ?? ?? ??           // if-ge ?, ?, :????
      44 ?? ?? ??           // aget ??, ??, ??
      B7 ??                 // xor-int/2addr (xor is fairly rare in legit code)
      71 ?? ?? ?? ?? ??     // invoke-static
      0A ??                 // move-result ??
      97 ?? ?? ??           // xor-int ??, ??, ??
      D8 00 00 01           // add-int/lit8 v0, v0, 0x1
      01 ??                 // move ?, ?
      28 F2                 // goto
      21 80                 // array-length
      D8 00 00 FE           // add-int/lit8 v0, v0, -0x2
      44 ?? ?? ??           // aget ??
      B7 ??                 // xor-int/2addr
      21 ??                 // invoke-static
      D8 ?? ?? ??           // add-int
      44 ?? ?? ??           // aget
      B7 ??                 // xor-int/2addr
    }
```

File: apkid/rules/apk/protectors.yara (hunk 6 above)
Comment (Fare9) [reply apkid/rules/apk/protectors.yara:102]: The application that @apkunpacker provided I think is an older version, after checking the diffing with the recent compiled versions I found that lacks some of the current protections, but I haven't found any versioning system, so probably the asset could be in newer versions, I wouldn't be 100% sure about it.

**Comment by enovella** [comment]
@Fare9 LGTM! Happy to merge when you reply my last comments.

File: apkid/rules/apk/protectors.yara (hunk 5 above)
Comment (enovella) [reply apkid/rules/apk/protectors.yara]: If you remove the file `assets/talsec`, will we have false positives on apps having these two native libraries?

File: apkid/rules/apk/protectors.yara (hunk 6 above)
Comment (enovella) [reply apkid/rules/apk/protectors.yara:102]: In this case, it is recommended to keep both rules (one with the `talsec` asset file and the one with decryption routine based on DEX opcodes.

**Comment by Fare9** [comment]
Okay, I will add a last commit, one with the asset the other without it, the rule will be _old and _new, and yes I will be happy to include the code from the offsets, in that way as you said is easier to understand the rule (I was worried that the code would be too verbose)

File: apkid/rules/dex/protectors.yara [hunk 8]
```diff
@@ -111,3 +111,68 @@ rule appdome_dex : protector
...
+  meta:
+    description = "FreeRASP"
+    sample      = "e10b8772fd9b6aaf8ba030c5bcb324fb9b91f34e893a62bdf238629df856e047"
+    url         = "https://www.talsec.app/freerasp-in-app-protection-security-talsec"
+    author      = "Fare9"
+  
+  strings:
+    // Decryption method found in DEX files, since strings will change
+    // and other offsets change, we add ?? to some instructions
+    $decryption = {
```
Comment (enovella) [inline apkid/rules/dex/protectors.yara:126]: Excellent! Thanks for clarifying. Now the comment above is not required

**Pull request merged**